Fix CI missing `uses` parameter in Run clippy

CPerezz opened the pull request:
In #38 an error was introduced in the CI so that it wasn't possible to
run them since they were panicking before starting.

- Edit the CI config and move clippy to a sepparated job.
- Edit the CI config and add global RUSTFLAGS env var.
- Edit the CI config and use better toolchain config.

Resolves: #39

Co-authored-by: Brandon H. Gomes <bhgomes@pm.me>

CPerezz committed:
Fix CI missing `uses` parameter in Run clippy

In #38 an error was introduced in the CI so that it wasn't possible to
run them since they were panicking before starting.

Adding the missing `uses` inside of the `job` should fix it.

Resolves: #39

bhgomes commented on .github/workflows/ci.yml:
```diff
@@ -27,6 +27,7 @@ jobs:
       - name: Run clippy
         with:
           command: clippy
+           uses: actions-rs/cargo@v1
```
Shouldn't this go above the `with` statement?

CPerezz replied:
You're completely right!! I force-pushed!

CPerezz committed:
Move clippy to a sepparate job

By doing that, we can overwrite the toolchain and use `stable` which
should always contain a working version of `clippy`.

CPerezz committed:
Remove redundant/duplicated imports

CPerezz committed:
Add RUSTFLAGS config global in CI

bhgomes commented on .github/workflows/ci.yml:
```diff
@@ -24,15 +24,13 @@ jobs:
-          args: --verbose --release --workspace --examples --tests --all-features -- -D warnings
       - name: Run tests
+        env: 
+          RUSTFLAGS: "-D warnings"
```
This should be enabled at the global `env` so that we have a more uniform calling convention.

CPerezz replied:
Ohh I did not see the other usages! My bad

bhgomes commented on .github/workflows/ci.yml:
```diff
@@ -98,3 +96,23 @@ jobs:
+        with:
+          toolchain: stable
+          override: true
+          components: clippy
```
I think this should be left as `override: false`. The point of keeping the library at `nightly` was to allow us to use unstable features if needed. If the CI is not using `nightly` then we cannot use it as a valid test, since it isn't using the same tools we are using.

If we use the `rust-toolchain.toml` file I suggested, then `actions-rs` should automatically download the right tools, and also it enforces that when developing, maintainers of this project have at the very least `rustfmt` and `clippy` installed, which are requirements of the CI for merging anyway.

bhgomes replied:
If we don't want to change the `rust-toolchain` then we should change the CI to this:
```yaml
with:
  override: false
  components: clippy
```

CPerezz replied:
Makes more sense. I agree.

Wasn't aware about these latest changes in the rust-toolchain file stuff!

Thanks for the update!

bhgomes replied:
Sorry, it looks like they didn't finish upgrading to the new version: https://github.com/actions-rs/toolchain/issues/126. I usually don't use `actions-rs`, I just do `rustup add component clippy`. So better to do the second option I outlined above.

CPerezz replied:
Exactly.  I was just going to say that. 
Seeing https://github.com/actions-rs/toolchain#the-toolchain-file seems that the naming convention differs. But supports the same date format.

bhgomes replied:
Yeah, so I would keep the `rust-toolchain` the same and just have `clippy` in the YAML as above.

CPerezz replied:
Yepp! Agree! Resetting!

bhgomes commented on src/constraint_system/composer.rs:
```diff
@@ -22,15 +22,13 @@ use crate::constraint_system::Variable;
 use crate::permutation::Permutation;
 use ark_ec::models::TEModelParameters;
 use ark_ec::PairingEngine;
+#[cfg(feature = "trace")]
 use ark_ff::{BigInteger, PrimeField};
```
I kept this on its own line since it uses a `feature` but I didn't check if we are doing this uniformly in the library, so either way is fine for now. Just letting you know what my preference is.

CPerezz replied:
Well, it's `rustfmt` the tool which merges the deps even if they have feature flags. 

Therefore I assume is expected to be on this way.

bhgomes replied:
Yeah, that's why I separate them by a newline so that `rustfmt` doesn't merge them. It's just a personal preference but let's keep it the way you have it now.

CPerezz replied:
Thanks! Let's see if we end up having more cases we will be able to judge better.
